### PR TITLE
fix: Throw error if tsconfig.json path is incorrect

### PR DESCRIPTION
**Pull request by harsha509**
## Proposed changes

Throw an error `Path to tsconfig.json is incorrect in config file`. if tsconfig.json path is incorrect

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### Reviewers: @webdriverio/project-committers

<a href="https://gitpod.io/#https://github.com/webdriverio/webdriverio/pull/8260"></a>

### CODE REVIEW

**Comment by harsha509** [comment]
Screenshot 2022-04-21 at 10 01 36 PM

File: packages/wdio-runner/src/index.ts [hunk 1]
```diff
@@ -98,6 +98,16 @@ export default class Runner extends EventEmitter {
         this._specs = specs
         this._caps = caps
 
+        /*
+        * Checks tsconfig.json path, throws error if it doesn't exist
+        */
+
+        if (args.autoCompileOpts?.tsNodeOpts?.project) {
+            if (!fs.existsSync(args.autoCompileOpts.tsNodeOpts.project)) {
+                throw new Error('Path to tsconfig.json is incorrect in config file')
```
Comment (praveendvd) [inline packages/wdio-runner/src/index.ts:107]: '''

Comment (praveendvd) [inline packages/wdio-runner/src/index.ts:107]: `throw new Error('Path to tsconfig is incorrect in config. File ${ args.autoCompileOpts.tsNodeOpts.project) not found`

As tsconfig.json can have another names should the error be more specific

File: packages/wdio-config/src/lib/ConfigParser.ts [hunk 2]
```diff
@@ -40,6 +42,16 @@ export default class ConfigParser {
     ) {}
 
     autoCompile() {
+        /**
+         * Checks tsconfig.json path, throws error if it doesn't exist
+         */
+        if (this._config.autoCompileOpts?.tsNodeOpts?.project) {
+            const tsconfigPath = path.resolve(this._config.autoCompileOpts.tsNodeOpts.project)
+            if (!fs.existsSync(tsconfigPath)) {
```
Comment (christian-bromann) [inline packages/wdio-config/src/lib/ConfigParser.ts:50]: Let's not use `existsSync` as it blocks the main thread and better practice here is to use `await fs.exists(...)` from `fs/promise` and make the method async

File: packages/wdio-config/src/lib/ConfigParser.ts [hunk 3]
```diff
@@ -1,5 +1,11 @@
 import merge from 'deepmerge'
 import logger from '@wdio/logger'
+import { exists } from 'fs'
+import { promisify } from 'util'
+const fs = {
+    exists: promisify(exists)
+}
```
Comment (christian-bromann) [inline packages/wdio-config/src/lib/ConfigParser.ts:7]: Let's use:

```suggestion
import fs from 'fs/promises'
```

and do it this way:

```ts
const isExisting = await fs.access(pathToDir).then(() => true, () => false)
```

File: packages/wdio-config/src/lib/ConfigParser.ts [hunk 4]
```diff
@@ -1,6 +1,8 @@
 import merge from 'deepmerge'
 import logger from '@wdio/logger'
+import path from 'path'
 import type { Capabilities, Options, Services } from '@wdio/types'
+import { canAccess } from '@wdio/utils'
```
Comment (christian-bromann) [inline packages/wdio-config/src/lib/ConfigParser.ts:5]: let's have function imports before type imports, so let's move it a line up

File: packages/wdio-config/src/lib/ConfigParser.ts [hunk 5]
```diff
@@ -1,6 +1,8 @@
 import merge from 'deepmerge'
 import logger from '@wdio/logger'
+import path from 'path'
```
Comment (christian-bromann) [inline packages/wdio-config/src/lib/ConfigParser.ts:3]: Let's import node modules before npm packages. I realise now we might want to eslint this.

File: packages/wdio-config/src/lib/ConfigParser.ts [hunk 6]
```diff
@@ -39,7 +41,17 @@ export default class ConfigParser {
...
     ) {}
 
-    autoCompile() {
+    async autoCompile() {
+        /**
+         * Checks tsconfig.json path, throws error if it doesn't exist
+         */
+        if (this._config.autoCompileOpts?.tsNodeOpts?.project) {
+            const tsconfigPath = path.resolve(this._config.autoCompileOpts.tsNodeOpts.project)
+            if (!canAccess(tsconfigPath)) {
```
Comment (christian-bromann) [inline packages/wdio-config/src/lib/ConfigParser.ts:50]:
```suggestion
            if (!(await canAccess(tsconfigPath))) {
```

Comment (harsha509) [reply packages/wdio-config/src/lib/ConfigParser.ts:50]: `canAccess` internally uses `fs.accessSync`. 
do we need await in here ?

Comment (christian-bromann) [reply packages/wdio-config/src/lib/ConfigParser.ts:51]: Ah! Then we don't, thanks for checking.

File: packages/wdio-config/src/lib/ConfigParser.ts [hunk 7]
```diff
@@ -39,7 +42,17 @@ export default class ConfigParser {
...
+                throw new Error(`Provided tsconfig file path '${tsconfigPath}' in wdio config is incorrect. Is it correctly set in wdio config ?`)
+            }
+        }
```
Comment (christian-bromann) [inline packages/wdio-config/src/lib/ConfigParser.ts:54]: Given that everything is sync here, I'ld suggest to move this into the [`loadTypeScriptCompiler`](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-config/src/utils.ts#L113-L133) function.

**Review by christian-bromann — COMMENTED** [review]
Once you moved the code, I recommend to add a test in `packages/wdio-config/tests/utils.test.ts` and check if the function throws and error if you pass in an invalid function. By passing in [mock parameters](https://jestjs.io/docs/mock-functions) you can also assert whether the right imports are being made. Have a look how Jest mocks work and when you write a test pass in an object that mimics the actual one. Also, don't worry about typing here, feel free to cast everything to any. I don't really see much value making sure that tests are type safe.

Let me know if you have any specific questions.

**Comment by christian-bromann** [comment]
@harsha509 mind rebasing on top of current `main` branch? Just pushed a [committ](https://github.com/webdriverio/webdriverio/pull/8303) that fixes some issues we had with TypeScript and testing.

File: packages/wdio-config/src/lib/ConfigParser.ts (hunk 7 above)
Comment (harsha509) [reply packages/wdio-config/src/lib/ConfigParser.ts:54]: Sure!

**Review by christian-bromann — CHANGES_REQUESTED** [review]
@harsha509 anything I can help with here?

I think what is missing is to move the if statement into the [`loadTypeScriptCompiler`](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-config/src/utils.ts#L113-L133) function as it makes more sense to check it there. Then a test would be great which should be fairly easy to add.

Let me know if I can help with anything.

**Comment by harsha509** [comment]
> @harsha509 anything I can help with here?
> 
> I think what is missing is to move the if statement into the [`loadTypeScriptCompiler`](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-config/src/utils.ts#L113-L133) function as it makes more sense to check it there. Then a test would be great which should be fairly easy to add.
> 
> Let me know if I can help with anything.

Sure Will update PR asap.

**Comment by harsha509** [comment]
PR is a bit stale!

Will re-raise clean PR!

**Pull request closed without merging**